editor/editor_command_palette.cpp

```diff
@@ -0,0 +1,254 @@
+		const String search_text = command_search_box->get_text();
+		const bool empty_search = search_text == "";
+
+		// Filter possible candidates.
+		Vector<Pair<String, String>> entries;
+		for (int i = 0; i < actions_list->size(); i++) {
+			String key_name = (*actions_list)[i];
+			String display_name = _get_command_name(key_name);
+			if (!empty_search && search_text.is_subsequence_ofi(display_name)) {
+				// To-do: Rank the Entries....
+				// r.score = empty_search ? 0 : _score_path(search_text, files[i].to_lower());
+				// print_line("found entry : " + (*actions_list)[i]);
+				Pair entry = Pair(key_name, display_name);
+				entries.push_back(entry);
+			}
+		}
+
+		actions_list->clear();
+
+		TreeItem *root = search_options->get_root();
+		root->clear_children();
+
+		if (entries.size() > 0) {
+			// if (!empty_search) {
+			// 	SortArray<Entry, EntryComparator> sorter;
+			// 	sorter.sort(entries.ptrw(), entries.size());
+			// }
+
+			const int entry_limit = MIN(entries.size(), 300);
+			for (int i = 0; i < entry_limit; i++) {
+				TreeItem *ti = search_options->create_item(root);
+				ti->set_tooltip(0, entries[i].first);
+				ti->set_text(0, entries[i].second);
+				// ti->set_icon(0, *icons.lookup_ptr(entries[i].path.get_extension()));
+			}
+
+			TreeItem *to_select = root->get_first_child();
+			to_select->select(0);
+			to_select->set_as_cursor(0);
+			search_options->scroll_to_item(to_select);
+
+			get_ok_button()->set_disabled(false);
+		} else {
+			search_options->deselect_all();
+
+			get_ok_button()->set_disabled(true);
+		}
+	} else {
+		print_error("Failed to fetch action_list.");
+	}
+}
+
+void EditorCommandPalette::_text_changed(const String &p_newtext) {
+	_update_command_search();
+}
+
+void EditorCommandPalette::_bind_methods() {
+	ADD_SIGNAL(MethodInfo("execute_command"));
+	ADD_SIGNAL(MethodInfo("open_resource"));
+}
+
+void EditorCommandPalette::_sbox_input(const Ref<InputEvent> &p_ie) {
+	Ref<InputEventKey> k = p_ie;
+	if (k.is_valid()) {
+		switch (k->get_keycode()) {
+			case KEY_UP:
+			case KEY_DOWN:
+			case KEY_PAGEUP:
+			case KEY_PAGEDOWN: {
+				search_options->call("_gui_input", k);
+				command_search_box->accept_event();
+			} break;
+		}
+	}
+}
+
+void EditorCommandPalette::_cleanup() {
+	print_line("cleaning everything."); //i think this is only for files part but should see.
+}
+
+void EditorCommandPalette::_confirmed() {
+	TreeItem *selected_option = search_options->get_selected();
+	if (selected_option == nullptr) {
+		print_error("Selected Null");
+		return;
+	}
+	set_selected_commmad(selected_option->get_tooltip(0));
+	_cleanup();
+	emit_signal("execute_command");
+	_hide_command_palette(); // window hide.
+}
+
+void EditorCommandPalette::_hide_command_palette() {
+	hide();
+}
+
+void EditorCommandPalette::_text_confirmed(const String &p_text) {
+	_confirmed();
+}
+
+String EditorCommandPalette::_get_command_name(const String &p_key) {
+	return commands[p_key].name;
```

Comment:
This does not need to b e a function either.

Reply:
yep